Update specification in line with data model definition design (#118)
ballerina-platform/persist-tools: docs/spec/spec.md

## Contents

1. [Overview](#1-overview)
2. [Initializing Persistence Layer in Bal Project](#2-initializing-the-bal-project-with-persistence-layer)
3. [Generating Persistence Derived Types and Clients](#3-generating-persistence-derived-types-and-clients)
4. [Push Persistence Schema to the Data Provider](#4-push-persistence-schema-to-the-data-provider)

## 1. Overview
This specification elaborates on the `Persist CLI Tool` commands.

## 2. Initializing the Bal Project with Persistence Layer

```bash
bal persist init
```

The command initializes the bal project with the persistence layer. This command includes the following steps,

1. Create persist directory
   This directory should contain all data model definition files. This file will define the required entities as per the [`persist` specification](https://github.com/ballerina-platform/module-ballerina-persist/blob/main/docs/spec/spec.md#2-data-model-definition)
2. Create a model definition file in persist directory
   It will create a file with same name as the package name if no files are present in the `persist` directory. It will create an empty file with required imports(`import ballerina/persist as _;`).

> 3. Create database_configurations.bal file inside the `default` module in generated directory

Reviewer: ```suggestion
3. Create database_configurations.bal file inside the `default` module in the generated directory if the definition filename inside the `persist` directory is the same as the Ballerina package name. If not, create the bal file inside the submodule named as the definition file name.
```

   This file contains configuration variables to initialize the data stores
    ```ballerina
    import ballerinax/mysql.driver as _;

    configurable int port = ?;
    configurable string host = ?;
    configurable string user = ?;
    configurable string database = ?;
    configurable string password = ?;
   ```
4. Update Ballerina.toml with database configurations.
   It will update the Ballerina.toml file with the configurations needed for the `bal persist push` command to work.
    ```ballerina

>     [persist.<package name>.storage.mysql]

Reviewer: ```suggestion
    [persist.<data_model_name>.storage.mysql]
```

    host = "localhost"
    port = 3306
    user = "root"
    password = ""
    database = "" 
   ```
5. Create(Update) Config.toml file inside the Ballerina project.
   It will create(update) `Config.toml` file with configurables used to initialize variables in Step 3.
    ```ballerina

>     [<package name>]

Reviewer: ```suggestion
    [<data_model_name>]
```

    host = "localhost"
    port = 3306
    user = "root"
    password = ""
    database = ""
    ```

The directory structure will be,
```
medical-center
├── generated
         └── database_configuration.bal
├── persist
         └── medical-center.bal
├── Ballerina.toml
├── Config.toml
└── main.bal
```

Behaviour of the `init` command,
- User should invoke the command within a bal project

> - If the user invokes the command twice, it will not fail. It will execute the steps once again.

Reviewer: ```suggestion
- If the user invokes the command twice, it will not fail. It will verify that all the configurations are in place for the definition files defined inside the `persist` directory. If not, add missing configurations and files.
```

## 3. Generating Persistence Derived Types and Clients

```bash
bal persist generate
```

The command will generate [Derived Entity Types and Persist Clients](https://github.com/ballerina-platform/module-ballerina-persist/blob/main/docs/spec/spec.md#3-derived-entity-types-and-persist-clients)
as per the `persist` specification.

It will add generated files under the conventions,
1. If the file name is the same as the package name, it will generate the files under the `default` module.
   ```
   medical-center
   ├── generated
         ├── database_configuration.bal
         ├── generated_client.bal
         └── generated_types.bal
   ├── persist
         └── medical-center.bal
   ├── Ballerina.toml
   ├── Config.toml
   └── main.bal
   ```

> 2. If the file name is different to the package name, it will generate the files under a new submodule with the same name as the file.

Reviewer: ```suggestion
2. If the file name is different from the package name, it will generate the files under a new submodule with the same name as the file.
```
